src/string/rabin-karp.md

```diff
@@ -7,16 +7,27 @@ e_maxx_link: rabin_karp
 # Rabin-Karp Algorithm for string matching
 
 This algorithm is based on the concept of hashing, so if you are not familiar with string hashing, refer to the [string hashing](string-hashing.md) article.
- 
 This algorithm was authored by Rabin and Karp in 1987.
 
-Problem: Given two strings - a pattern $s$ and a text $t$, determine if the pattern appears in the text and if it does, enumerate all its occurrences in $O(|s| + |t|)$ time.
+Problem: Given two strings - a length $m$ pattern to find in a length $n$ text, find all matches in $\Theta(m+n) = \Theta(n)$ average time.
```

Comment:
I don't think the randomization affects the correctness if you check the hash collisions. If you ignore them then yes.

Reply:
If the task is to find _all_ occurrences of the pattern in the text, you can't check them all, it would lead to worst-time $O(nm)$ even with a perfect hash function. The current implementation and the task statement heavily imply that we just report all hash coincidences as a match.

Reply:
I reported the worst-case time too. I do not think we should compromise the correctness because of the worst-case time which should be rare. CLRS has a little bit on it.

Reply:
Checking occurrences is somewhat fine if you only want to check that whether the pattern occurs at all (and possibly report one specific match). We can suggest doing a full match check in this specific case, but then should also provide some bounds on the probability that the number of false matches is high.

For the formulation of finding _all_ matches though we shouldn't assume that the number of valid matches is low at all, this is just a kind of assumptions that _will_ be wrong in competitive programming context with a reasonably well-prepared test set.